In [1]:
import pandas as pd
import numpy as np
import os

In [2]:
path = r'/Users/pgtips/Desktop/Instacart Basket Analysis/02.Data/Original Data'

# Products_df

In [3]:
df_prods = pd.read_csv(os.path.join(path,'products.csv'),index_col = False)

In [4]:
df_prods.shape

(49693, 5)

In [5]:
df_prods.head(1)

product_id                product_name  aisle_id  department_id  prices
0           1  Chocolate Sandwich Cookies        61             19     5.8

In [6]:
df_prods.dtypes

product_id         int64
product_name      object
aisle_id           int64
department_id      int64
prices           float64
dtype: object

##### We have to change the dtype of 'order_id' and 'user_id' as its a unique combination of numbers, to make sure that the system doesnt see it as numbers

In [7]:
df_prods['product_id'] = df_prods['product_id'].astype(str)
df_prods['aisle_id'] = df_prods['aisle_id'].astype(str)
df_prods['department_id'] = df_prods['department_id'].astype(str)

In [8]:
df_prods.dtypes

product_id        object
product_name      object
aisle_id          object
department_id     object
prices           float64
dtype: object

##### Since changing the dtype we can check for mixed dtypes with the code below

# Duplicates

In [9]:
for col in df_prods.columns.tolist():
  weird = (df_prods[[col]].applymap(type) != df_prods[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_prods[weird]) > 0:
    print (col)

product_name


##### Finding duplicates in product name

In [10]:
df_dups_prods = df_prods[df_prods.duplicated()]

In [11]:
df_dups_prods

product_id                                       product_name aisle_id  \
462          462                  Fiber 4g Gummy Dietary Supplement       70   
18459      18458                                         Ranger IPA       27   
26810      26808               Black House Coffee Roasty Stout Beer       27   
35309      35306  Gluten Free Organic Peanut Butter & Chocolate ...      121   
35495      35491                            Adore Forever Body Wash      127   

      department_id  prices  
462              11     4.8  
18459             5     9.2  
26810             5    13.4  
35309            14     6.8  
35495            11     9.9

In [12]:
df_prods.drop_duplicates()

product_id                                       product_name aisle_id  \
0              1                         Chocolate Sandwich Cookies       61   
1              2                                   All-Seasons Salt      104   
2              3               Robust Golden Unsweetened Oolong Tea       94   
3              4  Smart Ones Classic Favorites Mini Rigatoni Wit...       38   
4              5                          Green Chile Anytime Sauce        5   
...          ...                                                ...      ...   
49688      49684          Vodka, Triple Distilled, Twist of Vanilla      124   
49689      49685                 En Croute Roast Hazelnut Cranberry       42   
49690      49686                                   Artisan Baguette      112   
49691      49687         Smartblend Healthy Metabolism Dry Cat Food       41   
49692      49688                             Fresh Foaming Cleanser       73   

      department_id  prices  
0                19     5.8  
1                13     9.3  
2                 7     4.5  
3                 1    10.5  
4                13     4.3  
...             ...     ...  
49688             5     5.3  
49689             1     3.1  
49690             3     7.8  
49691             8     4.7  
49692            11    13.5  

[49688 rows x 5 columns]

##### We can see it was successful since the df_prods had (49693 rows) and now (49688 rows)

### Missing Values

In [13]:
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [14]:
df_prods_na = df_prods[df_prods['product_name'].isnull()==True]

In [15]:
df_prods_na.shape

(16, 5)

In [16]:
df_prods_na

product_id product_name aisle_id department_id  prices
33            34          NaN      121            14    12.2
68            69          NaN       26             7    11.8
115          116          NaN       93             3    10.8
261          262          NaN      110            13    12.1
525          525          NaN      109            11     1.2
1511        1511          NaN       84            16    14.3
1780        1780          NaN      126            11    12.3
2240        2240          NaN       52             1    14.2
2586        2586          NaN      104            13    12.4
3159        3159          NaN      126            11    13.1
3230        3230          NaN      120            16    14.4
3736        3736          NaN       41             8    14.8
4283        4283          NaN       77             7    14.4
4790        4790          NaN       91            16    14.5
38187      38183          NaN       39            12    20.9
40444      40440          NaN      120            16    14.8

In [17]:
df_prods.shape

(49693, 5)

In [18]:
df_prods = df_prods[df_prods['product_name'].isnull() == False]

In [19]:
df_prods.shape

(49677, 5)

##### Succesfully droped the 16 NaN columns

### Descriptive Stat

In [20]:
df_prods.describe().apply(lambda s: s.apply(lambda x: format(x, 'g')))

prices
count    49677
mean   9.99316
std    453.593
min          1
25%        4.1
50%        7.1
75%       11.1
max      99999

In [21]:
df_prods_outliner = df_prods[df_prods['prices'] >= 450]

In [22]:
df_prods_outliner

product_id                      product_name aisle_id department_id  \
21554      21553  Lowfat 2% Milkfat Cottage Cheese      108            16   
33666      33664             2 % Reduced Fat  Milk       84            16   

        prices  
21554  14900.0  
33666  99999.0

##### We detected two outliners

In [23]:
df_prods=df_prods[df_prods['prices']<=450]

In [24]:
df_prods.describe()

prices
count  49675.000000
mean       7.680552
std        4.199287
min        1.000000
25%        4.100000
50%        7.100000
75%       11.100000
max       25.000000

In [25]:
df=df_prods[df_prods['prices']>=20]

##### df is a dataframe with products that are more then 20$ for us to see what these rather expensiv items are

In [26]:
df.shape

(277, 5)

In [27]:
df.head(25)

product_id                                       product_name aisle_id  \
39           40  Beef Hot Links Beef Smoked Sausage With Chile ...      106   
83           84                                         Lamb Shank        7   
304         305                         Wild Forest Raw Neem Honey       29   
692         692                                        Skirt Steak      122   
907         907                               Premium Sliced Bacon      106   
1117       1117                                       Turkey Bacon      106   
1932       1932                             Soy Chorizo Vegetarian      106   
1992       1992                       Beef (101445) Summer Sausage      106   
2005       2005                         Fresh Ground Turkey Breast       35   
2496       2496                                          Bockwurst      106   
2503       2503                  Usda Prime Beef Top Sirloin Steak      122   
2735       2735                                Pork Loin Back Ribs        7   
2763       2763                       Uncured Hickory Smoked Bacon      106   
3078       3078                                      Chicken Kabob       35   
3161       3161                         Flake Style Imitation Crab       15   
3293       3293  Seasoned & Cooked Pulled White Meat Chicken wi...       49   
3529       3529                        Fully Cooked Original Bacon      106   
3775       3775                   Ham Steak, Boneless, Honey Cured        7   
4087       4087                 Boneless & Skinless Chicken Thighs       49   
4169       4169                              Original Pork Sausage      106   
4192       4192        Hardwood Smoked Peppered Thick Sliced Bacon      106   
4347       4347                     Extralean Ground Turkey Breast       49   
4421       4421                      Organic Beef Uncured Hot Dogs      106   
4922       4922                        Summer Sausage Beef Premium      106   
5215       5215                                    Beef Cube Steak      122   

     department_id  prices  
39              12    22.5  
83              12    24.3  
304             13    20.0  
692             12    20.7  
907             12    20.0  
1117            12    21.6  
1932            12    23.8  
1992            12    24.5  
2005            12    22.5  
2496            12    21.5  
2503            12    23.8  
2735            12    20.9  
2763            12    20.0  
3078            12    22.4  
3161            12    23.9  
3293            12    21.6  
3529            12    24.3  
3775            12    21.5  
4087            12    20.2  
4169            12    21.7  
4192            12    21.3  
4347            12    22.9  
4421            12    21.1  
4922            12    23.3  
5215            12    23.8

##### We wanted to check for further outliner but the products seems all to be premium porducts or rather expensive like Steaks etc.

In [28]:
df_prods.to_csv(r'/Users/pgtips/Desktop/Instacart Basket Analysis/05 Sent to Clint/Prepared Data/products.csv', index = False)

In [29]:
df_prods.shape

(49675, 5)